File: backend/app/domain/enums.py
"""冻结的枚举全集。

规格：docs/SPEC.md §2.1（枚举形状必须现在定死）、§5.3、§9.1、§9.2、§9.4、§12.3。

Gate-0 第 16 条要求「只产已声明枚举」：运行时产出的所有枚举值必须是本模块声明的
子集，由 tests/test_guards.py 的 `TestOnlyDeclaredEnums`（`-m enum_subset`）机械检查；
同一处还把 SPEC §2.1 冻结的词表逐字钉住，改词表必须是一次显式的规格修订。
因此本模块是唯一的枚举来源，任何地方不得使用裸字符串字面量代替这些枚举。
"""

from __future__ import annotations

from enum import StrEnum


class DocumentRole(StrEnum):
    """文档角色。SPEC §1.1。"""

    QUOTATION = "QUOTATION"
    PURCHASE_ORDER = "PURCHASE_ORDER"
    PROFORMA_INVOICE = "PROFORMA_INVOICE"


#: 单据在贸易链条上的先后顺序。chain_stage 的方向由它决定（SPEC §9.4）。
ROLE_ORDER: dict[DocumentRole, int] = {
    DocumentRole.QUOTATION: 0,
    DocumentRole.PURCHASE_ORDER: 1,
    DocumentRole.PROFORMA_INVOICE: 2,
}

#: role_signature（`Q1:P2:I0`）的角色缩写与固定顺序。SPEC §3.2。
#:
#: 生产方（`MatchGroupDraft.role_signature`）与消费方（报告 / 前端的中文化）
#: 必须读同一份定义：形状是「每个角色恰好一段、顺序固定」，
#: 少一段就不是本系统产出的签名，翻译它等于替用户断言「这个角色不参与」。
SIGNATURE_TAGS: tuple[tuple[str, DocumentRole], ...] = (
    ("Q", DocumentRole.QUOTATION),
    ("P", DocumentRole.PURCHASE_ORDER),
    ("I", DocumentRole.PROFORMA_INVOICE),
)


class ParseStatus(StrEnum):
    """SPEC §5.3 冻结词表。"""

    PENDING = "PENDING"
    OK = "OK"
    NEEDS_REVIEW = "NEEDS_REVIEW"
    REJECTED = "REJECTED"
    FAILED = "FAILED"


class ParseReasonCode(StrEnum):
    """结构化拒绝/告警原因。

    必须是字段而非拼在中文文案里，否则 golden test 只能靠字符串匹配（SPEC §5.3）。
    """

    UNSUPPORTED_EXT = "UNSUPPORTED_EXT"
    ENCRYPTED = "ENCRYPTED"
    CORRUPT = "CORRUPT"
    INSUFFICIENT_TEXT = "INSUFFICIENT_TEXT"
    UNSUPPORTED_TEXT_LAYER = "UNSUPPORTED_TEXT_LAYER"
    ROW_LIMIT = "ROW_LIMIT"
    SHEET_LIMIT = "SHEET_LIMIT"
    NO_TABLE_FOUND = "NO_TABLE_FOUND"
    FORMULA_WITHOUT_CACHE = "FORMULA_WITHOUT_CACHE"
    FILE_TOO_LARGE = "FILE_TOO_LARGE"


class Scope(StrEnum):
    """差异作用域。SPEC §9。"""

    DOCUMENT = "DOCUMENT"
    LINE_ITEM = "LINE_ITEM"
    CALCULATION = "CALCULATION"


class SubjectKind(StrEnum):
    """difference_key 中 subject 的类别。SPEC §3.3。"""

    DOCUMENT_ROLE = "DOCUMENT_ROLE"
    MATCH_GROUP = "MATCH_GROUP"


class DifferenceType(StrEnum):
    """SPEC §9.1 全集（原计划 7 个 + INCOMPARABLE）。"""

    VALUE_CONFLICT = "VALUE_CONFLICT"
    MISSING_VALUE = "MISSING_VALUE"
    CALCULATION_ERROR = "CALCULATION_ERROR"
    UNMATCHED_LINE_ITEM = "UNMATCHED_LINE_ITEM"
    AMBIGUOUS_MATCH = "AMBIGUOUS_MATCH"
    SEMANTIC_DIFFERENCE = "SEMANTIC_DIFFERENCE"
    EXTRACTION_UNCERTAIN = "EXTRACTION_UNCERTAIN"
    #: 结构上无法比较（币种混合、单位换算未知、计价基准不同）。
    #: 绝不能坍缩成 VALUE_CONFLICT（假警报）或不产出（危险的沉默）。SPEC §9.1。
    INCOMPARABLE = "INCOMPARABLE"


class Severity(StrEnum):
    """风险等级。INFO 是本次修订新增（SPEC §9.4）。

    没有 INFO，「买方砍价」这类正常业务只能落在 REVIEW 及以上，仍会污染总览。
    """

    CRITICAL = "CRITICAL"
    WARNING = "WARNING"
    REVIEW = "REVIEW"
    INFO = "INFO"


#: 排序用（数字越小越严重）。对外输出的稳定排序依赖它。
SEVERITY_ORDER: dict[Severity, int] = {
    Severity.CRITICAL: 0,
    Severity.WARNING: 1,
    Severity.REVIEW: 2,
    Severity.INFO: 3,
}


class ChainStage(StrEnum):
    """比较发生在贸易链条的哪一段。SPEC §9.4。

    严重度按链路阶段查表，不按字段——这是本次修订在领域层面最重要的一条。
    """

    OFFER_TO_ORDER = "OFFER_TO_ORDER"  # Q  -> PO
    ORDER_TO_CONFIRMATION = "ORDER_TO_CONFIRMATION"  # PO -> PI
    OFFER_TO_CONFIRMATION = "OFFER_TO_CONFIRMATION"  # Q  -> PI
    WITHIN_DOCUMENT = "WITHIN_DOCUMENT"


def chain_stage_for(baseline: DocumentRole, target: DocumentRole) -> ChainStage:
    """由一对角色推出 chain_stage。参数顺序无关（内部按 ROLE_ORDER 定向）。"""
    lo, hi = sorted((baseline, target), key=lambda r: ROLE_ORDER[r])
    if lo is DocumentRole.QUOTATION and hi is DocumentRole.PURCHASE_ORDER:
        return ChainStage.OFFER_TO_ORDER
    if lo is DocumentRole.PURCHASE_ORDER and hi is DocumentRole.PROFORMA_INVOICE:
        return ChainStage.ORDER_TO_CONFIRMATION
    if lo is DocumentRole.QUOTATION and hi is DocumentRole.PROFORMA_INVOICE:
        return ChainStage.OFFER_TO_CONFIRMATION
    raise ValueError(f"同角色不构成链路阶段: {baseline} / {target}")


class Verdict(StrEnum):
    """四态比较器（+MISSING）。SPEC §9.2。"""

    EQUAL = "EQUAL"
    DIFFERENT = "DIFFERENT"
    INCOMPARABLE = "INCOMPARABLE"
    UNCERTAIN = "UNCERTAIN"
    MISSING = "MISSING"


class MatchMethod(StrEnum):
    """SPEC §3.2 match_group.match_method。"""

    SKU_EXACT = "SKU_EXACT"
    CUSTOMER_PART_MAP = "CUSTOMER_PART_MAP"  # MVP-1
    FUZZY_CANDIDATE = "FUZZY_CANDIDATE"  # MVP-1
    UNMATCHED = "UNMATCHED"
    USER_MANUAL = "USER_MANUAL"  # MVP-1


class MultiplicityState(StrEnum):
    """组内每个角色是否唯一。MULTI 阻断字段比较（SPEC §9.5）。"""

    UNIQUE_PER_ROLE = "UNIQUE_PER_ROLE"
    MULTI_PER_ROLE = "MULTI_PER_ROLE"


class CoverageState(StrEnum):
    """组覆盖了几个角色。缺口不阻断比较（SPEC §9.5）。"""

    FULL = "FULL"
    PARTIAL = "PARTIAL"
    ISOLATED = "ISOLATED"


class SelectionState(StrEnum):
    AUTO_SELECTED = "AUTO_SELECTED"
    CANDIDATE = "CANDIDATE"  # MVP-1
    USER_SELECTED = "USER_SELECTED"  # MVP-1
    USER_REJECTED = "USER_REJECTED"  # MVP-1


class MatchGroupStatus(StrEnum):
    RESOLVED = "RESOLVED"
    NEEDS_USER_DECISION = "NEEDS_USER_DECISION"


class ReviewStatus(StrEnum):
    """SPEC §12.3 冻结词表。"""

    OPEN = "OPEN"
    CONFIRMED_DIFFERENCE = "CONFIRMED_DIFFERENCE"
    ACCEPTED_DIFFERENCE = "ACCEPTED_DIFFERENCE"
    NEEDS_CONFIRMATION = "NEEDS_CONFIRMATION"
    IGNORED = "IGNORED"
    RESOLVED = "RESOLVED"


class IdentityStrength(StrEnum):
    """WEAK 的差异重跑后不继承审核状态——宁可可见地丢，不可错挂。SPEC §3.3。"""

    STRONG = "STRONG"
    WEAK = "WEAK"


class ExtractionMethod(StrEnum):
    """SPEC §6.3 优先级：user_confirmed > alias > layout。

    白名单同时是「LLM 没有偷偷变成必需依赖」的机器证明（SPEC §14）。
    """

    ALIAS = "alias"
    LAYOUT = "layout"
    USER_CONFIRMED = "user_confirmed"


#: Gate-0 断言用：MVP-0 允许出现的提取方法。出现白名单外的值即失败。
ALLOWED_EXTRACTION_METHODS: frozenset[str] = frozenset(
    {ExtractionMethod.ALIAS, ExtractionMethod.LAYOUT, ExtractionMethod.USER_CONFIRMED}
)


class ValueKind(StrEnum):
    """FieldSpec.value_kind。SPEC §4。"""

    TEXT = "TEXT"
    DECIMAL = "DECIMAL"
    MONEY = "MONEY"
    CURRENCY = "CURRENCY"
    DATE = "DATE"
    ENUM = "ENUM"
    STRUCTURED = "STRUCTURED"


class ValueSource(StrEnum):
    """values_by_document 里每个值的来源。SPEC §9.6。

    报告要发给老板和客户，「这个数是机器读的还是人填的」必须精确到字段。
    """

    PARSER = "PARSER"
    USER_CORRECTION = "USER_CORRECTION"


class CorrectionKind(StrEnum):
    OVERRIDE = "OVERRIDE"
    CONFIRM = "CONFIRM"


class EvidenceSourceType(StrEnum):
    XLSX_CELL = "XLSX_CELL"
    XLSX_RANGE = "XLSX_RANGE"
    PDF_TEXT = "PDF_TEXT"  # MVP-1
    DERIVED = "DERIVED"


class ProjectStatus(StrEnum):
    DRAFT = "DRAFT"
    READY = "READY"
    COMPARED = "COMPARED"

File: frontend/src/labels/enums.ts
/**
 * 「英文枚举 -> 中文」映射表（硬约束 #1：界面中文单语，禁止 i18n 框架）。
 *
 * 与 backend/app/exports/html.py 的标签表逐条对齐，报告和界面上同一个枚举
 * 必须是同一个中文词——业务员对着报告核对界面时词不一样会当成两件事。
 *
 * 所有查表函数遇到**未登记的标识符一律回退成原始标识符**，绝不抛异常：
 * 显示一个英文词很难看，但静默丢一条差异是本产品最危险的失败方式。
 */

function lookup(table: Readonly<Record<string, string>>, key: string | null | undefined): string {
  if (key === null || key === undefined || key === '') return '—'
  return table[key] ?? key
}

export const ROLE_LABEL: Readonly<Record<string, string>> = {
  QUOTATION: '报价单',
  PURCHASE_ORDER: '采购订单',
  PROFORMA_INVOICE: '形式发票',
}

export function roleLabel(role: string | null | undefined): string {
  return lookup(ROLE_LABEL, role)
}

/** 角色在贸易链条上的顺序，用于稳定排序（报价单 -> 采购订单 -> 形式发票）。 */
export const ROLE_ORDER: Readonly<Record<string, number>> = {
  QUOTATION: 0,
  PURCHASE_ORDER: 1,
  PROFORMA_INVOICE: 2,
}

export function roleRank(role: string): number {
  return ROLE_ORDER[role] ?? 99
}

export const SEVERITY_LABEL: Readonly<Record<string, string>> = {
  CRITICAL: '严重',
  WARNING: '警告',
  REVIEW: '待复核',
  INFO: '提示',
}

export function severityLabel(severity: string): string {
  return lookup(SEVERITY_LABEL, severity)
}

/**
 * 严重度**不靠颜色单独承载**（考虑色觉障碍）：形状记号 + 中文标签 + 颜色三重冗余。
 * 整页去掉颜色后依然可读。
 */
export const SEVERITY_MARK: Readonly<Record<string, string>> = {
  CRITICAL: '▲',
  WARNING: '◆',
  REVIEW: '■',
  INFO: '●',
}

export function severityMark(severity: string): string {
  return SEVERITY_MARK[severity] ?? '○'
}

/** CSS 修饰类后缀。未知严重度落到 unknown，不会顶掉样式。 */
export function severitySlug(severity: string): string {
  const known: Readonly<Record<string, string>> = {
    CRITICAL: 'critical',
    WARNING: 'warning',
    REVIEW: 'review',
    INFO: 'info',
  }
  return known[severity] ?? 'unknown'
}

export const SEVERITY_RANK: Readonly<Record<string, number>> = {
  CRITICAL: 0,
  WARNING: 1,
  REVIEW: 2,
  INFO: 3,
}

export function severityRank(severity: string): number {
  return SEVERITY_RANK[severity] ?? 99
}

export const DIFFERENCE_TYPE_LABEL: Readonly<Record<string, string>> = {
  VALUE_CONFLICT: '取值冲突',
  MISSING_VALUE: '字段缺失',
  CALCULATION_ERROR: '金额计算不符',
  UNMATCHED_LINE_ITEM: '行项目未对齐',
  AMBIGUOUS_MATCH: '匹配存在歧义',
  SEMANTIC_DIFFERENCE: '表述差异',
  EXTRACTION_UNCERTAIN: '提取不确定',
  INCOMPARABLE: '无法比较',
}

export function differenceTypeLabel(value: string): string {
  return lookup(DIFFERENCE_TYPE_LABEL, value)
}

export const SCOPE_LABEL: Readonly<Record<string, string>> = {
  DOCUMENT: '文档级',
  LINE_ITEM: '行项目',
  CALCULATION: '单据内计算',
}

export function scopeLabel(value: string): string {
  return lookup(SCOPE_LABEL, value)
}

export const CHAIN_STAGE_LABEL: Readonly<Record<string, string>> = {
  OFFER_TO_ORDER: '报价单 → 采购订单',
  ORDER_TO_CONFIRMATION: '采购订单 → 形式发票',
  OFFER_TO_CONFIRMATION: '报价单 → 形式发票',
  WITHIN_DOCUMENT: '单份单据内部',
}

export function chainStageLabel(value: string): string {
  return lookup(CHAIN_STAGE_LABEL, value)
}

export const VALUE_SOURCE_LABEL: Readonly<Record<string, string>> = {
  PARSER: '机器读取',
  USER_CORRECTION: '人工修正',
}

export function valueSourceLabel(value: string): string {
  return lookup(VALUE_SOURCE_LABEL, value)
}

export const EVIDENCE_SOURCE_LABEL: Readonly<Record<string, string>> = {
  XLSX_CELL: 'Excel 单元格',
  XLSX_RANGE: 'Excel 区域',
  PDF_TEXT: 'PDF 文本',
  DERIVED: '推导（算式）',
}

export function evidenceSourceLabel(value: string): string {
  return lookup(EVIDENCE_SOURCE_LABEL, value)
}

export const PARSE_STATUS_LABEL: Readonly<Record<string, string>> = {
  PENDING: '待处理',
  OK: '解析成功',
  NEEDS_REVIEW: '需人工复核',
  REJECTED: '已拒绝',
  FAILED: '解析失败',
}

export function parseStatusLabel(value: string): string {
  return lookup(PARSE_STATUS_LABEL, value)
}

export const PARSE_REASON_LABEL: Readonly<Record<string, string>> = {
  UNSUPPORTED_EXT: '不支持的文件类型',
  ENCRYPTED: '文件已加密',
  CORRUPT: '文件损坏',
  INSUFFICIENT_TEXT: '可用文本过少',
  UNSUPPORTED_TEXT_LAYER: '文本层不可用',
  ROW_LIMIT: '超出行数上限',
  SHEET_LIMIT: '超出工作表数上限',
  NO_TABLE_FOUND: '未定位到订单表格',
  FORMULA_WITHOUT_CACHE: '公式没有缓存值',
  FILE_TOO_LARGE: '文件过大',
}

export function parseReasonLabel(value: string | null): string {
  return lookup(PARSE_REASON_LABEL, value)
}

export const REVIEW_STATUS_LABEL: Readonly<Record<string, string>> = {
  OPEN: '未处理',
  CONFIRMED_DIFFERENCE: '确认存在差异',
  ACCEPTED_DIFFERENCE: '接受该差异',
  NEEDS_CONFIRMATION: '待重新确认',
  IGNORED: '忽略',
  RESOLVED: '已解决',
}

export function reviewStatusLabel(value: string): string {
  return lookup(REVIEW_STATUS_LABEL, value)
}

export const PROJECT_STATUS_LABEL: Readonly<Record<string, string>> = {
  DRAFT: '草稿',
  READY: '待检查',
  COMPARED: '已检查',
}

export function projectStatusLabel(value: string): string {
  return lookup(PROJECT_STATUS_LABEL, value)
}

export const IDENTITY_STRENGTH_LABEL: Readonly<Record<string, string>> = {
  STRONG: '强身份',
  WEAK: '弱身份',
}

export function identityStrengthLabel(value: string): string {
  return lookup(IDENTITY_STRENGTH_LABEL, value)
}

/** 字段 key -> 中文名。与 backend/app/domain/fields.py 的 label_zh 一致。 */
export const FIELD_LABEL: Readonly<Record<string, string>> = {
  internal_sku: '内部型号',
  customer_part_number: '客户料号',
  description: '产品描述',
  specification: '规格',
  color: '颜色',
  quantity: '数量',
  unit: '单位',
  unit_price: '单价',
  currency: '币种',
  line_total: '行金额',
  packaging_quantity: '装箱数量',
  carton_count: '箱数',
  remarks: '备注',
  document_number: '单据号',
  document_date: '单据日期',
  buyer_name: '买方',
  seller_name: '卖方',
  incoterm: '贸易术语',
  incoterm_named_place: '贸易术语地点',
  incoterm_version: '贸易术语版本',
  payment_terms: '付款条件',
  delivery_terms: '交期',
  destination: '目的地',
  shipping_method: '运输方式',
  grand_total: '总金额',
}

export function fieldLabel(value: string | null): string {
  return lookup(FIELD_LABEL, value)
}

/** 没有取到值时的占位。空白在表格里必须**看得见**。 */
export const EMPTY_VALUE_TEXT = '（未提取到）'

/** SPEC §1.3：集合驱动比较引入的新失效模式，不得隐藏。逐字与报告一致。 */
export const COVERAGE_WARNING = '缺席角色 = 未检查，不等于无差异'

/** SPEC §1.2「不是什么」。必须常驻可见。 */
export const DISCLAIMER = '本工具只能辅助核对，不判断哪份文件正确，不构成贸易、法律或财务结论。'

/**
 * 删除项目的措辞。CLAUDE.md 安全红线：只能说「删除数据库记录与磁盘文件」，
 * **不得宣称「安全擦除」「不可恢复」**。
 */
export const DELETE_PROJECT_WARNING = '将删除数据库记录与磁盘上的原始文件。'

export function formatFileSize(bytes: number): string {
  if (bytes < 1024) return `${bytes} B`
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`
  return `${(bytes / (1024 * 1024)).toFixed(1)} MB`
}

/** 后端返回 ISO 字符串；无效值原样显示，不做时区推断。 */
export function formatDateTime(value: string | null): string {
  if (!value) return '—'
  return value.replace('T', ' ').slice(0, 19)
}
